contrib/babelfishpg_tsql/src/catalog.c

```diff
@@ -2799,6 +2828,614 @@ rename_procfunc_update_bbf_catalog(RenameStmt *stmt)
+	else
+	{
+		/*
+		 * In case of REVOKE, we remove the new privilege and keep the previous privilege as it is.
+		 */
+		current_permission = old_permission & ~new_permission;
+	}
+
+	if (current_permission == 0)
+	{
+		remove_entry_from_bbf_schema_perms(schema_name, object_name, grantee, object_type);
+		return;
+	}
+
+	bbf_schema_rel = table_open(get_bbf_schema_perms_oid(),
+							RowExclusiveLock);
+
+	ScanKeyInit(&scanKey[0],
+				Anum_bbf_schema_perms_dbid,
+				BTEqualStrategyNumber, F_INT2EQ,
+				Int16GetDatum(dbid));
+	ScanKeyEntryInitialize(&scanKey[1], 0,
+				Anum_bbf_schema_perms_schema_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(schema_name));
+	ScanKeyEntryInitialize(&scanKey[2], 0,
+				Anum_bbf_schema_perms_object_name,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(object_name));
+	ScanKeyInit(&scanKey[3],
+				Anum_bbf_schema_perms_permission,
+				BTEqualStrategyNumber, F_INT4EQ,
+				Int32GetDatum(old_permission));
+	ScanKeyEntryInitialize(&scanKey[4], 0,
+				Anum_bbf_schema_perms_grantee,
+				BTEqualStrategyNumber,
+				InvalidOid,
+				tsql_get_server_collation_oid_internal(false),
+				F_NAMEEQ,
+				CStringGetDatum(grantee));
+
+	scan = table_beginscan_catalog(bbf_schema_rel, 5, scanKey);
+	tuple_bbf_schema = heap_getnext(scan, ForwardScanDirection);
+	if (HeapTupleIsValid(tuple_bbf_schema))
+	{
+		bbf_schema_dsc = RelationGetDescr(bbf_schema_rel);
+		/* Build a tuple to insert */
+		MemSet(new_record_bbf_schema, 0, sizeof(new_record_bbf_schema));
+		MemSet(new_record_nulls_bbf_schema, false, sizeof(new_record_nulls_bbf_schema));
+		MemSet(new_record_repl_bbf_schema, false, sizeof(new_record_repl_bbf_schema));
+
+		new_record_bbf_schema[BBF_SCHEMA_PERMS_PERMISSION] = Int32GetDatum(current_permission);
+		new_record_repl_bbf_schema[BBF_SCHEMA_PERMS_PERMISSION] = true;
+
+		new_tuple = heap_modify_tuple(tuple_bbf_schema,
+									bbf_schema_dsc,
+									new_record_bbf_schema,
+									new_record_nulls_bbf_schema,
+									new_record_repl_bbf_schema);
+
+		CatalogTupleUpdate(bbf_schema_rel, &new_tuple->t_self, new_tuple);
+		heap_freetuple(new_tuple);
+	}
+
+	table_endscan(scan);
+	table_close(bbf_schema_rel, RowExclusiveLock);
+}
+
+/*
+ * Checks if a particular catalog entry exists in BABELFISH_SCHEMA_PERMISSIONS.
+ */
+bool
+entry_exists_in_bbf_schema(const char *schema_name,
+							const char *object_name,
+							const char *grantee)
+{
+	Relation	bbf_schema_rel;
+	HeapTuple	tuple_bbf_schema;
+	ScanKeyData	scanKey[4];
+	SysScanDesc	scan;
+	bool	catalog_entry_exists = false;
+	int16	dbid = get_cur_db_id();
+
+	/* Immediately return false, if grantee is NULL or PUBLIC. */
+	if ((grantee == NULL) || (strcmp(grantee, "public") == 0))
+		return false;
+
+	/* Immediately return false, if SCHEMA name is NULL. */
+	if (schema_name == NULL)
```

Comment:
We must throw errors/return false for system/shared schemas

Reply:
Will do.